### DISCUSSION TITLE
This junction object is difficult to configure

### DISCUSSION THREAD

**Discussion by toando23** [deejay-hub/timeline-lwc#221] [Q&A] [answered]
Hello all I'm 8 hours into trying to plot one relationship and I'm calling in reinforcements.  Thanks in advance for the assistance.  

This is error I am getting: 
<img width="1887" alt="image" src="https://user-images.githubusercontent.com/115432073/194782476-aacafafd-bad5-4314-844b-bb5976bf7309.png">

Here is what I am trying to do:

I have a Person org.  Accounts are both traditional accounts and people accounts.
Accounts own patients (animals).
Patients have events that create orders.  I can see orders in the related field of patients when they are created.

Accounts are 1:many with patients.  Patients have a master-detail relationship with accounts.

On Accounts, I can successfully see both patients (DOB) and Orders (date of entry) on the timeline.

However,
When I go into patients to create the timeline for orders I get the "unexpected token" error. I suspect this is the junction issue in the config file.  Here is a screenshot of the .mdt file

<img width="1116" alt="image" src="https://user-images.githubusercontent.com/115432073/194782755-354f9a1b-70ed-4e1e-9c48-5777cb28d71a.png">

I've tried every which way including changing the lightning page to parent record from "Use this patient" to account owner.  When I change it to account owner it shows all orders and all patients on a selected patient (not ideal).

<img width="1605" alt="image" src="https://user-images.githubusercontent.com/115432073/194782902-32efc153-43fc-403d-afc3-fb4dc618b003.png">

**Comment by deejay-hub** [comment, marked as answer]
Reinforcements have arrived. The config guide is available [here](https://quip.com/ahFCA9VBEKtr) that discusses some nuances but since you may have already looked at it we'll look at your specific example.

Let me just understand this a little better.
1. You have a business account (standard Account) linked to a custom object called Patient__c via a master detail relationship. So each patient (animal) is linked to one parent Account but each Account can have one or more Patients.

So when you view the custom object Patient__c you see a master detail relationship like the one below.
<img width="1360" alt="image" src="https://user-images.githubusercontent.com/7021883/194784459-dd62a975-36fd-412f-b7fd-637bcf1a6d3c.png">

2. You are using the standard Orders object linked to a Patient__c via a lookup relationship. So each Order is linked to one Patient but each Patient can have one or more Orders.

So when you view the Order object you see a lookup relationship to Patient__c like the one below?

<img width="1355" alt="image" src="https://user-images.githubusercontent.com/7021883/194784560-ad5e961e-5a35-4863-97ba-ab4bc82a18af.png">

Let's assume I have understood correctly. Add look at the steps needed to plot Orders on the Patient object.

Step 1: Identify the relationship name that defines how the parent and child Objects are linked. In the above use case I want to plot Orders on the Patient object so I drill down on the field in step 2 above.

<img width="1393" alt="image" src="https://user-images.githubusercontent.com/7021883/194786059-3e8c75cd-97c9-46d9-89a4-ac743befd390.png">

Step 2: Make a note of the value in the *Child Relationship Name* field. In the above screen you can see *Orders* but note that for custom relationships your value might be different.

Step 3. Decide where you want to plot Orders in the timeline. This is typically a field on the child object itself. Looks like you wanted the Order Start Date which maps to the API name *EffectiveDate*

Step 4: Decide which field value to appear when we plot on the timeline. Let's pick the Order Number in this example - you can change it if you like. Again this is typically a field on the child object.

Step 5: Create a new Timeline_Config__mdt record.

<img width="1395" alt="image" src="https://user-images.githubusercontent.com/7021883/194786268-f11e4f65-6140-4963-a5ce-449ae4c58544.png">

Then refresh the browser. Logging out and back in in necessary. To see Orders related to Patients by dragging the Time Warp component into the Patient page layout.

<img width="1406" alt="image" src="https://user-images.githubusercontent.com/7021883/194786314-c7103a1c-4565-4f10-bb4c-f0d2d4989657.png">

**Reply by toando23** [reply]
Yes you are correct on the first 2 assumptions.  

1. Patient has a master-detail relationship:
<img width="1542" alt="image" src="https://user-images.githubusercontent.com/115432073/194787732-2866b4c1-db78-4a65-9cd4-162074ef1ed5.png">

2. Order has a look up to patient__C 
<img width="1411" alt="image" src="https://user-images.githubusercontent.com/115432073/194787762-37a6a1e4-5891-4e92-b050-24381d355102.png">

The next step is where we differ but let me explain why: When I start a new order, user types in account and then ONLY patients of that account show up.  It presents a child being assigned to the wrong account and in the long run an account paying for a child they do not own.
<img width="1641" alt="image" src="https://user-images.githubusercontent.com/115432073/194787973-6faae124-8e4d-4413-a287-0ef53d037aeb.png">

Now in your examples the Child relationship name was Orders.  My was patients.  I tried patients__r and got same error.  I then tried orders__r.  Same result which is:

Could it be a package error I'm getting since it says: Setup error?

<img width="1875" alt="image" src="https://user-images.githubusercontent.com/115432073/194788491-a5a1e57d-38b7-4a53-899a-f85b79b6bf73.png">

**Reply by toando23** [reply]
I figured it out!  I used a lowercase p in patients__r!1 Works now!

<img width="1871" alt="image" src="https://user-images.githubusercontent.com/115432073/194788822-4b47a4c2-c3a2-411b-8f5a-d322da39dd58.png">

**Reply by deejay-hub** [reply]
Excellent news. Glad you got it working.

**Comment by toando23** [comment]
@deejay-hub I've created a monster!  When I showed my test users what was accomplished they asked for more!  And I didn't think it was going to be a big ask given that I had already completed the tasks for parent/child objects. 

But first a background to remind you:


1. PersonAccount Org.
2. Accounts have 1:many to patients
3. Patients have 1:many to orders

I have successfully shown on a timeline Account with patients AND orders AND Events.
<img width="1877" alt="image" src="https://user-images.githubusercontent.com/115432073/196042228-8e4ffbd3-b780-4b77-a080-a490193ef241.png">

I have successfully shown patients and only their patient's orders, and Events.
<img width="1879" alt="image" src="https://user-images.githubusercontent.com/115432073/196042264-bd2f722b-3ff0-47f3-978a-23ab077f4493.png">


And now the problem statement.  I also have contacts which are not Accounts.  The contact roles are secondary owner, trainer, and referring vet.  These are look-ups to the Patient__c object.  For these contacts I can show the patients associated to the contact and by the different icon, the contact role. Below is an example of where somebody is a secondary owner of Lucky, and the referring vet for Kelly Cat2.

<img width="1867" alt="image" src="https://user-images.githubusercontent.com/115432073/196042923-47e8c902-94db-4eb6-b370-b7f50a6ebace.png">

What I cannot get to populate are the orders associated to Lucky and Cat2 for the various contacts.  I know this is because there is no lookup to secondary, trainer, and referring vet on the Order__c object.  But I also read I can use dot notation where a contact and lookup patient, and then display Order__c.

Here is metadata mapping that works to show secondary owner and patients.  
<img width="1616" alt="image" src="https://user-images.githubusercontent.com/115432073/196043290-86af9f95-0756-4f34-bd92-a6b249d0a857.png">

Here is the metadata that works to show patients to orders.
<img width="1619" alt="image" src="https://user-images.githubusercontent.com/115432073/196043325-76baec9a-ab7b-4f9a-b200-29fba3885fd1.png">

Here is where I am struggling with dot notation and the linkage to show Orders via the Patient__c using the existing patients__r relationship between Order and Patient.

<img width="1620" alt="image" src="https://user-images.githubusercontent.com/115432073/196043612-9645da75-d1aa-433f-991f-36652516fe23.png">

What am I missing?

**Comment by deejay-hub** [comment]
It's hard to tell from your description what your setup is. But let me give you a good rule of thumb. When deciding if it's possible to plot an object...navigate to the page layout of the parent and look for a related list that corresponds to the child object you want to plot. If you find one...Time Warp can plot it. If not it can't. Dot notation is for traversing field values like a formula fields.

I suspect your Parent does not have a direct relationship to the object you're trying to plot. Since the same vet and trainer can be on multiple Patients it's hard to see what you're trying to accomplish.

(The only exception is the Account to Task/Event/Email and Call. Indirect relationships are supported here by virtue of the fact that Salesforce will populate Open Activities and Activity History objects with grandchild records.)

**Reply by toando23** [reply]
Well I guess if that's the limitations than I won't feel defeated anymore.  Your suspicion is correct, there is not a direct relationship between the contacts and the orders.  The only relationship is patients have a 1:many relationship with orders, and Accounts and contacts have a many:1 relationship with the patient (meaning you might have different humans that play a role in the patient).

I was hoping to be able to map the relationship where patient acts almost as a junction between the contacts and the orders.  Thank you for your time.  Awesome product.
